### PR TITLE
[FIXED JENKINS-24824] Collect all asynchronously deleted directories

### PR DESCRIPTION
**Pull request by olivergondza**
https://issues.jenkins-ci.org/browse/JENKINS-24824

@jenkinsci/code-reviewers: This is the first client of https://github.com/jenkinsci/resource-disposer-plugin. New utility plugin to track resources that require some time / more retries to delete and reports failure to do so. That is what I would like you to review, in fact.

### CODE REVIEW

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java [hunk 1]
```diff
@@ -61,23 +69,46 @@ protected void perform(FilePath workspace) throws IOException, InterruptedExcept
                     new Object[] { workspace.getRemote(), deleteMe.getName() }
             );
             workspace.deleteRecursive();
+        } else {
+            String slaveName = toComputer(deleteMe.getChannel()).getName();
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: Missing null check

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java [hunk 2]
```diff
@@ -61,23 +69,46 @@ protected void perform(FilePath workspace) throws IOException, InterruptedExcept
...
+        private final String path;
+
+        private DisposableImpl(String node, String path) {
+            this.node = node;
+            this.path = path;
+        }
+
+        @Nonnull public State dispose() throws Exception {
+            Jenkins j = Jenkins.getInstance();
+            if (j == null) return State.DISPOSE; // Going down?
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: Maybe this state needs a better name

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java [hunk 3]
```diff
@@ -61,23 +69,46 @@ protected void perform(FilePath workspace) throws IOException, InterruptedExcept
...
+            ws.deleteRecursive();
+
+            return ws.exists()
+                ? State.DISPOSE // Failed to delete silently
+                : State.PURGED
+            ;
+        }
+
+        @Nonnull public String getDisplayName() {
+            return "Workspace " + (node.isEmpty() ? "master" : node) + ':' + path;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: NIT: missing localization support

**Comment by oleg-nenashev** [comment]
This PR would cause collisions with https://github.com/jenkinsci/ws-cleanup-plugin/pull/27. but it should be easy to fix them

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java (hunk 2 above)
Comment (olivergondza) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: `DISPOSE` meas the resource is yet to be disposed. Any suggestions?

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java (hunk 3 above)
Comment (olivergondza) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: Right, though not a conceptual problem with resource-disposer.

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java (hunk 1 above)
Comment (olivergondza) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: The only situation I can imagine when `FilePath` representing workspace in a build step is backed up by a channel that does not belong to any computer is that the computer got disconnected in the middle of the build. Or perhaps someone is reusing these classes for something nasty. Anyway, getting back to synchronous deletion will either work or fail with an exception resembling the cause. Will fix.

File: src/main/java/hudson/plugins/ws_cleanup/Wipeout.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/ws_cleanup/Wipeout.java]: Maybe NOT_DISPOSED or TO_BE_DISPOSED?

**Comment by oleg-nenashev** [comment]
Needs fix to make it mergeable again.
In general 👍 (excepting the enum value name)

**Comment by olivergondza** [comment]
Ok, I managed to get the test working again.

@tcollignon, @Jimilian, Review and testing appreciated.

**Comment by Jimilian** [comment]
@olivergondza Can you summarise, please, what we should to test? :)

**Comment by olivergondza** [comment]
Right, the async cleanup was moved to dedicated plugin that will retry until it succeeds surviving Jenkins restarts. Failure to delete so, even temporary ones, will be reported to instance admins as admin monitor.

**Comment by Jimilian** [comment]
Ok, we are going to rollout this version during weekend and test for a next week.

**Pull request merged**